doc/man/src/install-opencpi.1.txt
INSTALL-OPENCPI(1)
================


NAME
----
install-opencpi - command-line tool for installing the OpenCPI framework


SYNOPSIS
--------
*`./scripts/install-opencpi.sh`* ['<options>']


DESCRIPTION
-----------
The *`install-opencpi`* tool builds/installs the OpenCPI framework on the local
RCC development platform and runs a series of tests to verify the installation.
The tool is intended to be run from the top level (root directory) of the OpenCPI
source tree, and will emit an error message if run from any other directory.
Upon successful completion, the OpenCPI framework is ready for use in developing
and/or running various OpenCPI components and applications.

OPTIONS
-------
*`--minimal`*::
    Performs a standard installation except the tests are not run. 
    Suppressing the generation and running of tests saves time in cases where framework
    testing is obviously unnecessary, e.g., reinstalling the same version of the framework on a host.

*`--artifacts-only`*::
    Cleans intermediate build files when building workers and assemblies, keeps only artifacts 
    required for building at the next level or deploying onto hardware.

*`--no-kernel`*::
    Prevents building of the OpenCPI Linux kernel driver module. See the link:ocpidriver.1.html[ocpidriver(1)]
    man page and the 'OpenCPI User Guide' for details on when this driver module is required.

EXAMPLES
--------
. Install OpenCPI from the root directory of the OpenCPI source tree.
+
----------
./scripts/install-opencpi.sh
----------
. Install OpenCPI without running the installation tests.
+
----------
./scripts/install-opencpi.sh --minimal
----------
. Install OpenCPI without running the installation tests and without building the OpenCPI Linux kernel driver module.
+
----------
./scripts/install-opencpi.sh --minimal --no-kernel
----------

BUGS
----
See https://www.opencpi.org/report-defects

RESOURCES
---------
See the 'OpenCPI Installation Guide':
https://opencpi.gitlab.io/releases/latest/docs/OpenCPI_Installation_Guide.pdf

See the 'OpenCPI User Guide':
https://opencpi.gitlab.io/releases/latest/docs/OpenCPI_User_Guide.pdf

SEE ALSO
--------
link:ocpidev.1.html[ocpidev(1)]
link:ocpidriver.1.html[ocpidriver(1)]
link:ocpiview.1.html[ocpiview(1)]
link:ocpirun.1.html[ocpirun(1)]


COPYING
-------
Copyright \(C) 2023 OpenCPI www.opencpi.org. OpenCPI is free software:
you can redistribute it and/or modify it under the terms of the
GNU Lesser General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your
option) any later version.


CHANGELOG.md
# [v2.4.7](https://gitlab.com/opencpi/opencpi/-/compare/v2.4.6...v2.4.7) (2024-01-04)

Changes/additions since [OpenCPI Release v2.4.6](https://gitlab.com/opencpi/opencpi/-/releases/v2.4.6)

### Features
- **app**: updated the fsk_dig_radio_ctrl image. (!1057)(b6fcacfb)
- **comp**: support include and xi:include. (!995)(ebd35042)
- **comp**: replace xi:include in property files with include. (!995)(ebd35042)
- **comp,tests**: add ability to compare top N bits of fractional time in ocpi_testing. (!1144)(45d4d09e)
- **devops**: enable autoscaling for CI/CD pipelines. (!1115)(2b0e5ab6)
- **devops**: enable parallel building of workers in CI/CD pipelines. (!1115)(2b0e5ab6)
- **devops**: allow for skipping of certain CI/CD pipeline jobs if relevant changes are not made to the repo. (!1120)(5c7c286a)
- **devops**: make non-MR CI/CD pipelines start automatically after being launched. (!1124)(2da5d1c8)
- **devops**: remove sudo from centos7 docker images. (!1125)(048f09e4)
- **devops**: update ubuntu dockerfiles. (!1126)(8f2019c3)
- **devops**: update rocky8 dockerfile. (!1128)(6b9a432d)
- **devops**: allow projects using downstream-ci.yml to pass the `CI_OCPI_PLATFORMS` and `CI_OCPI_PROJECTS` variables when set in the project settings. (!1130)(baad9f06)
- **devops**: re-implement CI/CD pipeline HWIL. (!1137)(92efb17b)
- **devops**: enabling CI/CD pipelines to deploy full docker images to public registry. (!1141)(8c4d366d)
- **devops**: add Ubuntu18.04 support to CI pipelines. (!842)(2fc0efd5)
- **devops**: don't run/build unit tests in CI pipelines for platforms when there is not a worker of the appropriate model available. (!859)(1d5b13b7)
- **devops**: allow multiple devices of same platform in HWIL CI pipelines. (!862)(1332223f)
- **devops**: enable zcu102, picozed, microzed platforms in HWIL CI pipelines. (!862)(1332223f)
- **devops**: add source_sink application to CI/CD pipelines. (!867)(0eae0e52)
- **devops**: add testbias app to CI pipelines. (!875)(13b4d819)
- **devops**: prevent attempts to build inactive platforms in CI pipelines. (!875)(13b4d819)
- **devops**: let runner dictate device to connect to in HWIL CI pipelines. (!875)(13b4d819)
- **devops**: make pipeline ignore warnings from ocpidev. (!880)(ad9c8552)
- **devops**: add FSK app to CI pipelines. (!893)(feccd726)
- **devops**: add failed job report to CI pipelines. (!894)(c0026da7)
- **devops**: auto push develop and release docker images to gitlab container registry. (!954)(da29f8d2)
- **devops**: add zcu111 to pipeline. (!964)(7b6738bf)
- **devops**: fix section that pushes to gitlab registry to use predefined ci variables. (!985)(46482d35)
- **devops**: add x310 and x3u platforms to CI config to enable them to run properly in pipelines. (!1022)(243d0bbd)
- **devops**: update docs to support building docs in forked projects. (!1091)(80eba8b2)
- **devops,tools**: add `rocky8` RCC development platform support. (!874)(cd558c26)
- **devops,tools**: add `ubuntu22_04` RCC development platform support. (!885)(4f8b666c)
- **doc**: add install-opencpi man page. (!1021)(d179174a)
- **doc**: add ocpidev asset doc system usage information to relevant man pages. (!1023)(cb54577f)
- **doc**: fix ordering of options in man pages to be short form first. (!1023)(cb54577f)
- **doc**: add ethernet transport performance tuning guide. (!1033)(2f8d0d14)
- **doc**: rename HDL and RCC worker RST files to latest naming scheme. (!1035)(9d2a079f)
- **doc**: update component RST docs to use latest terminology and template format. (!1035)(9d2a079f)
- **doc**: update component RST block diagrams (*.odg, *.svg) to be exemplar. (!1035)(9d2a079f)
- **doc**: update RST files and directories for platform project device workers to use current organization, structure, and naming scheme. (!1036)(594fc672)
- **doc**: remove deprecated RST files and directories for platform project device workers. (!1036)(594fc672)
- **doc**: add node graph to Tutorial 1. (!1039)(123ce17f)
- **doc**: add node graph to tutorial 2. (!1040)(d85c577b)
- **doc**: move picoevb and zed RST files and images to be co-located with their assets. (!1041)(95806072)
- **doc**: update picoevb and zed GSG to refer to information in User Guide and Installation Guide. (!1041)(95806072)
- **doc**: add missing hyphen to `ocpidev` command for #3647. (!1142)(11896d00)
- **doc**: update tutorial docs for new code, and grammar issue #3646. (!1142)(11896d00)
- **doc,tools**: remove `--rcc-platform` from `ocpidev run` man page, for #3662. (!1142)(11896d00)
- **doc,tools**: add `ocpi_documentation_include` directive to ocpidoc. (!882)(877eb994)
- **doc,tools**: add RST templates for OSP-related documentation. (!1008)(82b22c6a)
- **hdl base**: add DG-RDMA Primitive Library. (!947)(7de0ace9)
- **hdl base**: add (optional) fairer arbitration scheme to SDP. (!947)(7de0ace9)
- **hdl base**: add DG-RDMA configuration device worker and proxy. (!947)(7de0ace9)
- **hdl base**: add Ethernet MAC dependency. (!947)(7de0ace9)
- **osp**: add Riviera-Pro support. (!906)(5e074965)
- **osp**: add zed_ether platform as an HDL DG-RDMA reference OSP. (!947)(7de0ace9)
- **runtime**: add ability to dump data when ocpirun receives CTL+C. (!1114)(6a5039d1)
- **runtime**: adjust error message and fix memory issue. (!872)(9fc49553)
- **runtime**: add worker property settings to system.xml, either app or artifact. (!877)(10c53c07)
- **runtime**: add statically defined devices in system.xml, with options to include static devices or not. (!877)(10c53c07)
- **runtime**: add various fixes to make DGRDMA work properly. (!877)(10c53c07)
- **runtime**: introduce loadtime workers which have artifact load/unload lifecycle, not app lifecycle. (!877)(10c53c07)
- **runtime**: add static discovery method and support for JTAG programming for Ethernet devices. (!947)(7de0ace9)
- **runtime**: improvements to datagram transfer driver. (!947)(7de0ace9)
- **runtime**: add new datagram2 transfer driver. (!947)(7de0ace9)

scripts/install-opencpi.sh
#!/bin/bash --noprofile
# This file is protected by Copyright. Please refer to the COPYRIGHT file
# distributed with this source distribution.
#
# This file is part of OpenCPI <http://www.opencpi.org>
#
# OpenCPI is free software: you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# OpenCPI is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

# Just check if it looks like we are in the source tree.
[ -d runtime -a -d build -a -d scripts -a -d tools ] || {
  echo "Error:  this script ($0) is not being run from the top level of the OpenCPI source tree."
  exit 1
}
set -e

#
# Quick and dirty argument parsing:
# getopt(s) overhead not required.
#
PARAMS=""

while (( "$#" )); do
  case "$1" in
    -d|--distro)
      export OCPI_DISTRO_BUILD=1
      shift
      ;;
    --minimal)
      minimal=1
      shift
      ;;
    --artifacts-only)
      export OCPI_ARTIFACTS_ONLY=1
      shift
      ;;
    --optimize)
      optimize=1
      shift
      ;;
    --no-kernel)
      nokernel=1 # use ${nokernel:+whatever}
      shift
      ;;
    --use-env)
      useenv=1
      shift
      ;;
    -*) # unsupported flags
      echo "Error: Unsupported flag $1" >&2
      exit 1
      ;;
    *) # preserve positional arguments
      PARAMS="$PARAMS \"$1\""
      shift
      ;;
  esac
done

# set positional arguments in their proper place
eval set -- "$PARAMS"
unset PARAMS

# We do some bootstrapping here (that is also done in the scripts we call), in order to know
# the platform we are building

# Ensure exports (or cdk) exists and has scripts
echo "Initializing OpenCPI CDK" in $(pwd -P)
source ./scripts/init-opencpi.sh

#
# At this point, if the first positional parameter is either not
# set or is set and NULL, we may safely assume this is a framework
# installation on a local RCC development platform.  In a probably
# futile attempt to protect users from themselves, try to unset
# "OCPI_*" environment variables that may negatively affect the
# installation process.
#
if [ -z "$1" ]; then
  echo "Target platform not specified: assuming framework installation on local RCC development platform"
  echo "Cleaning OpenCPI environment..."
  source ./cdk/opencpi-setup.sh -i -v
fi

#
# If "--use-env" is specified, the OpenCPI environment is assumed
# to be set up properly, i.e., do not mess with it by calling the
# bootstrap script.  "ocpiadmin" installing an RCC platform is the
# expected use case, i.e., "ocpiadmin" calls this script with the
# "--use-env" flag.
#
if [ -z "$useenv" ]; then
  # Ensure CDK and TOOL variables
  OCPI_BOOTSTRAP="$(pwd)/cdk/scripts/ocpibootstrap.sh"; source "$OCPI_BOOTSTRAP"

  #
  # Because "ocpibootstrap.sh" is a special case in that it cannot contain any
  # bashisms or be called with arguments (per the comments at the beginning of
  # that script), OpenCPI environment fixup for the optimized framework install
  # case must be done afterward.  Problem is "ocpibootstrap.sh" sources
  # "opencpi-setup.sh -r", and while "opencpi-setup.sh" supports the "--optimize"
  # flag, cannot pass it into "ocpibootstrap.sh" :-(.
  #
  if [ -n "$optimize" ]; then
    source $OCPI_CDK_DIR/opencpi-setup.sh -r --optimize
  fi
fi

source $OCPI_CDK_DIR/scripts/util.sh

# This will set OCPI_TARGET_* vars.
echo -n "Finding target platform ... "
source "$OCPI_CDK_DIR/scripts/ocpitarget.sh" "$1" -
if [ -z "$OCPI_TARGET_PLATFORM_DIR" ]; then
  echo "Cannot find platform '$1'"
  exit 1
fi

echo "Found '$OCPI_TARGET_PLATFORM' ($OCPI_TARGET_PLATFORM_DIR)"

[ -n "$1" ] && shift
if [ -n "$OCPI_TARGET_PLATFORM" -a "$OCPI_TOOL_PLATFORM" != "$OCPI_TARGET_PLATFORM" ]; then
  echo "Assuming development host platform $OCPI_TOOL_PLATFORM is already installed."
fi

#
# Allow this to build for platforms defined in the inactive project or in OSPs
# that are not already registered.
#
[ -z "$OCPI_PROJECT_PATH" ] && export OCPI_PROJECT_PATH="$(pwd)/projects/inactive"
registry=$(getProjectRegistryDir)
for i in $(shopt -s nullglob; echo projects/osps/*); do
  direct=$(cd $i; pwd -P)
  for p in $(shopt -s nullglob; echo $registry/*); do
    [ "$direct" = "$(cd $p; pwd -P)" ] && echo OSP $i already registered && continue 2
  done
  [ -d "$i/rcc/platforms" ] && OCPI_PROJECT_PATH="$OCPI_PROJECT_PATH:$(pwd)/$i"
done

#
# Make sure we are running in a mode without any platform exports since this installation phase
# should *not* depend on platform exports.  This essentially unexports the platform.
#
OCPI_TARGET_PLATFORM_DIR="${OCPI_TARGET_PLATFORM_DIR%/lib}"
rm -r -f "${OCPI_TARGET_PLATFORM_DIR:?}/lib"

# We assume that install-packages.sh and install-prerequisites.sh do NOT depend on the platform's exports
./build/install-packages.sh "$OCPI_TARGET_PLATFORM"

# Export RCC platform
$OCPI_CDK_DIR/scripts/enable-rcc-platform.sh "$OCPI_TARGET_PLATFORM"

# Now build/install prerequisites.
./build/install-prerequisites.sh "$OCPI_TARGET_DIR"

# Build framework and built-in projects for target platform
eval $* ./build/build-opencpi.sh ${minimal:+--minimal} ${nokernel:+--no-kernel} "$OCPI_TARGET_DIR"

# Run RCC unit tests
if [ -n "$OCPI_TARGET_PLATFORM" -a "$OCPI_TOOL_PLATFORM" != "$OCPI_TARGET_PLATFORM" ]; then
  echo "When building/installing for cross-compiled platform $OCPI_TARGET_DIR, we are skipping tests."
elif [ -n "$minimal" ]; then
  echo "Avoiding installation tests since the --minimal option was given."
  echo 'Installation tests may be run at any time using the "ocpitest --no-hdl" command.'
else
  eval $* ./scripts/test-opencpi.sh --no-hdl ${nokernel:+--no-kernel}
fi
